### PR TITLE
[saiport.h]Support max accumulative headroom size[saiport] Add SAI_PORT_QOS_ATTR_MAXIMUM_HEADROOM_SIZE

### PR DESCRIPTION
**Pull request by stephenxs**
Support max accumulative headroom size, which is a read-only attribute.
For example, if a port has 2 PGs and headroom size of each is N bytes, the accumulative headroom size should be 2*N.
The motivation is that:
- We're going to have a port's headroom size automatically calculated according to its cable length and speed. If a user configured a very long cable length the headroom size can be very large as well, which will exceed the limit of switch chip and thus failing the SDK and then the orchagent.
- To prevent that case, we need to know the limit in hardware and prevent a limit-exceeding value from programming to switch chip by SAI.
- This attribute stands for that limit.

### CODE REVIEW

File: inc/saiport.h [hunk 1]
```diff
@@ -368,6 +368,14 @@ typedef enum _sai_port_attr_t
      */
     SAI_PORT_ATTR_QOS_SCHEDULER_GROUP_LIST,
 
+    /**
+     * @brief The maximum accumulative headroom size of a port in bytes
+     *
+     * @type sai_uint32_t
+     * @flags READ_ONLY
+     */
+    SAI_PORT_ATTR_QOS_MAXIMUM_ACCUMULATIVE_HEADROOM_SIZE,
```
Comment (rck-innovium) [inline inc/saiport.h:377]: 1) As discussed, please add comments to say this is applicable only for per-port per-PG headroom model (i.e., SAI_BUFFER_POOL_ATTR_XOFF_SIZE = 0)

2) Consider renaming the attribute.
SAI_PORT_ATTR_BUFFER_MAX_HEADROOM_SIZE

Comment (stephenxs) [reply inc/saiport.h]: Fixed.

**Comment by rlhui** [comment]
@rck-innovium , @JaiOCP  - would you please help review/approve this PR? Thanks.

File: inc/saiport.h [hunk 2]
```diff
@@ -368,6 +368,17 @@ typedef enum _sai_port_attr_t
      */
     SAI_PORT_ATTR_QOS_SCHEDULER_GROUP_LIST,
 
+    /**
+     * @brief The maximum accumulative headroom size of a port in bytes
```
Comment (rck-innovium) [inline inc/saiport.h:372]: 1) This line should be reworded to make it easier to understand. Like:

```
The sum of the headroom size of the ingress priority groups belonging to this port should not exceed SAI_PORT_ATTR_QOS_MAXIMUM_HEADROOM_SIZE value.
```

File: inc/saiport.h [hunk 3]
```diff
@@ -368,6 +368,17 @@ typedef enum _sai_port_attr_t
...
+     *
+     * This attribute is applicable only for per-port, per-PG headroom model
+     * (which means SAI_BUFFER_POOL_ATTR_XOFF_SIZE is zero)
+     *
+     * @type sai_uint32_t
+     * @flags READ_ONLY
+     */
```
Comment (rck-innovium) [inline inc/saiport.h:379]: What should be the value returned by the platforms that do not have such a limit?

File: inc/saiport.h (hunk 2 above)
Comment (stephenxs) [reply inc/saiport.h]: ok. Fixed.

File: inc/saiport.h (hunk 3 above)
Comment (stephenxs) [reply inc/saiport.h:380]: They can just return 0.

Comment (rck-innovium) [reply inc/saiport.h:380]: Thanks. Please update this in the comments as well so that the NOS/application handles the special case of 0.

Comment (stephenxs) [reply inc/saiport.h:382]: Done.

**Comment by stephenxs** [comment]
Hi @rlhui @JaiOCP ,
Could please you review/approve this PR?
Thanks.
Stephen

**Review by kcudnik — APPROVED** [review]

**Review by JaiOCP — APPROVED** [review]

**Comment by stephenxs** [comment]
Hi @rlhui 
as all the reviewers approved the PR, can we merge it?
thanks

**Pull request merged**